### PR TITLE
[tests] Refactor MongoDB.Driver tests to use testcontainers

### PR DESCRIPTION
**Pull request by radical**
###### Microsoft Reviewers: [Open in CodeFlow](https://microsoft.github.io/open-pr/?codeflow=https://github.com/dotnet/aspire/pull/2718)

### CODE REVIEW

File: eng/Versions.props [hunk 1]
```diff
@@ -46,5 +46,6 @@
     <MicrosoftEntityFrameworkCoreSqlServerPackageVersion>8.0.2</MicrosoftEntityFrameworkCoreSqlServerPackageVersion>
     <MicrosoftEntityFrameworkCoreToolsPackageVersion>8.0.2</MicrosoftEntityFrameworkCoreToolsPackageVersion>
     <MicrosoftNETRuntimeWorkloadTestingInternalVersion>9.0.0-preview.3.24151.4</MicrosoftNETRuntimeWorkloadTestingInternalVersion>
+    <TestcontainersPackageVersion>3.7.0</TestcontainersPackageVersion>
```
Comment (eerhardt) [inline eng/Versions.props:49]: Let's leave `eng\Versions.props` only have the versions that are managed by darc in it. Can you move this shared property to the top of `Directory.Packages.props`? I dislike having to look in multiple places. And when we get dependabot working again, I'm not sure it knows how to update `eng\Versions.props` files.

File: tests/Shared/RequiresDockerTheoryAttribute.cs [hunk 2]
```diff
@@ -3,12 +3,15 @@
...
+namespace Aspire.InternalTesting;
 
 // TODO: remove these attributes when Helix has a Windows agent with Docker support
 public class RequiresDockerTheoryAttribute : TheoryAttribute
 {
-    public static bool IsSupported => !OperatingSystem.IsWindows() || Environment.GetEnvironmentVariable("HELIX_WORKITEM_ROOT") is null;
+    // Not available on windows/CI
+    public static bool IsSupported => !OperatingSystem.IsWindows() ||
+                                      (Environment.GetEnvironmentVariable("HELIX_WORKITEM_ROOT") is null
+                                      && Environment.GetEnvironmentVariable("BUILD_BUILDID") is null);
```
Comment (eerhardt) [inline tests/Shared/RequiresDockerTheoryAttribute.cs:14]: Why do we need to add `BUILD_BUILDID`? 

I assume you are trying to support the case where these tests try to run on the azdo build pipeline machines, but does Docker work on the azdo pipeline machines on Linux?

File: tests/Aspire.RabbitMQ.Client.Tests/Aspire.RabbitMQ.Client.Tests.csproj [hunk 3]
```diff
@@ -6,6 +6,7 @@
   </PropertyGroup>
 
   <ItemGroup>
+    <Compile Include="..\Shared\RequiresDockerTheoryAttribute.cs" />
```
Comment (eerhardt) [inline tests/Aspire.RabbitMQ.Client.Tests/Aspire.RabbitMQ.Client.Tests.csproj:9]: We have a shared test library. Let's move this there instead so all these projects don't need to compile include the file. You can also put it in an existing namespace like `Aspire` or `Aspire.Components`, so all the files don't need to include a new one-off namespace.

File: tests/Aspire.MongoDB.Driver.Tests/ConformanceTests.cs [hunk 4]
```diff
@@ -92,6 +104,11 @@ protected override void TriggerActivity(IMongoClient service)
         service.ListDatabases(source.Token);
     }
 
+    protected override HealthStatus GetHealthStatus()
+        => _containerFixture.Container?.Health == TestcontainersHealthStatus.Healthy
+            ? HealthStatus.Healthy
+            : HealthStatus.Unhealthy;
```
Comment (eerhardt) [inline tests/Aspire.MongoDB.Driver.Tests/ConformanceTests.cs:110]: We should always be expecting that we can connect to the server now and that it is healthy. We shouldn't need this new virtual/override method. We just need to `override bool CanConnectToServer => RequiresDockerTheoryAttribute.IsSupported;`, which will make the `HealthCheckReportsExpectedStatus` test work correctly.

File: tests/Aspire.MongoDB.Driver.Tests/AspireMongoDBDriverExtensionsTests.cs [hunk 5]
```diff
@@ -10,13 +11,21 @@
...
+    public AspireMongoDBDriverExtensionsTests(MongoDbContainerFixture containerFixture)
+    {
+        _containerFixture = containerFixture;
+    }
+
+    private string DefaultConnectionString => _containerFixture.GetConnectionString();
+
     [Theory]
-    [InlineData("mongodb://localhost:27017/mydatabase", true)]
+    [InlineData("mongodb://localhost:2017/mydatabase", true)]
```
Comment (eerhardt) [inline tests/Aspire.MongoDB.Driver.Tests/AspireMongoDBDriverExtensionsTests.cs:28]: Why this change?

Comment (radical) [reply tests/Aspire.MongoDB.Driver.Tests/AspireMongoDBDriverExtensionsTests.cs:28]: debugging slipped through!

File: tests/Shared/RequiresDockerTheoryAttribute.cs (hunk 2 above)
Comment (radical) [reply tests/Shared/RequiresDockerTheoryAttribute.cs:14]: This was to make sure that we don't run any `RequiresDocker*` tests on windows/CI at all - helix or build machines.

File: tests/Aspire.RabbitMQ.Client.Tests/Aspire.RabbitMQ.Client.Tests.csproj (hunk 3 above)
Comment (radical) [reply tests/Aspire.RabbitMQ.Client.Tests/Aspire.RabbitMQ.Client.Tests.csproj:9]: Where is this shared test library? I don't see a csproj in `tests/Shared`.

Comment (eerhardt) [reply tests/Aspire.RabbitMQ.Client.Tests/Aspire.RabbitMQ.Client.Tests.csproj:9]: https://github.com/dotnet/aspire/tree/main/tests/Aspire.Components.Common.Tests

File: tests/Aspire.MongoDB.Driver.Tests/ConformanceTests.cs (hunk 4 above)
Comment (radical) [reply tests/Aspire.MongoDB.Driver.Tests/ConformanceTests.cs:110]: It was failing because the it got `Unhealthy` from the service.

https://github.com/radical/aspire/blob/46bce9b06b4e6ea10e3a68879bad53d24cb9c2a8/tests/Aspire.Components.Common.Tests/ConformanceTests.cs#L317-L320

Comment (eerhardt) [reply tests/Aspire.MongoDB.Driver.Tests/ConformanceTests.cs:110]: If the service says it is unhealthy, then the test should fail. We don't want to mask that problem. We want to know when something goes wrong.

Comment (radical) [reply tests/Aspire.MongoDB.Driver.Tests/ConformanceTests.cs:110]: Right, and this would get the real status of the container to match against what we get from the health check. I'll try to see why/when it is showing unhealthy.

**Comment by radical** [comment]
I'm not sure if removing `test_db` is the correct thing. Or if need to prime the db after creation, and before the tests are run.
 [`308535c` (#2718)](https://github.com/dotnet/aspire/pull/2718/commits/308535cdc28cc86ed77838e8189d8765a54be47c)

File: tests/Aspire.MongoDB.Driver.Tests/MongoDbContainerFixture.cs [hunk 6]
```diff
@@ -0,0 +1,49 @@
...
+            var mongoClient = new MongoClient(Container.GetConnectionString());
+            var createUserCommand = new BsonDocumentCommand<BsonDocument>(BsonDocument.Parse("""
+            {
+               createUser: "mongo",
+               pwd: "mongo",
+               roles: [ { role: 'readWrite', db: 'test_db' } ]
+            }
+            """));
+            await mongoClient.GetDatabase("test_db")
+                .RunCommandAsync(createUserCommand);
```
Comment (eerhardt) [inline tests/Aspire.MongoDB.Driver.Tests/MongoDbContainerFixture.cs:38]: In our "playground" app, we just do:

https://github.com/dotnet/aspire/blob/ded1eaef1592000b7c99136d1cc9a9085170f07b/playground/mongo/Mongo.ApiService/Program.cs#L19-L20

Is that enough? That seems simpler than what we have here.

File: tests/Aspire.MongoDB.Driver.Tests/MongoDbContainerFixture.cs [hunk 7]
```diff
@@ -0,0 +1,49 @@
...
+
+    public string GetConnectionString() => Container?.GetConnectionString() ??
+        throw new InvalidOperationException("The test container was not initialized.");
+
+    public async Task InitializeAsync()
+    {
+        if (RequiresDockerTheoryAttribute.IsSupported)
+        {
+            Container = new MongoDbBuilder()
+                .WithImage("mongo:7.0.5")
```
Comment (eerhardt) [inline tests/Aspire.MongoDB.Driver.Tests/MongoDbContainerFixture.cs:24]: I wonder if we should make it so these versions stay in sync between our tests and the AppHost:

https://github.com/dotnet/aspire/blob/ded1eaef1592000b7c99136d1cc9a9085170f07b/src/Aspire.Hosting/MongoDB/MongoDBBuilderExtensions.cs#L31

File: tests/Aspire.MongoDB.Driver.Tests/MongoDbContainerFixture.cs (hunk 6 above)
Comment (radical) [reply tests/Aspire.MongoDB.Driver.Tests/MongoDbContainerFixture.cs:38]: With the only the two lines above, it still fails with:
`,"msg":"Failed to authenticate","attr":{"client":"172.17.0.1:58864","isSpeculative":false,"isClusterMember":false,"mechanism":"SCRAM-SHA-1","user":"mongo","db":"test_db","error":"UserNotFound: Could not find user \"mongo\" for db \"test_db\"",`

The testcontainers use `mongo:mongo` by default.
How is the container setup/started in case of the playground?

File: tests/Aspire.MongoDB.Driver.Tests/MongoDbContainerFixture.cs (hunk 7 above)
Comment (radical) [reply tests/Aspire.MongoDB.Driver.Tests/MongoDbContainerFixture.cs:24]: Yes, definitely!

File: tests/Aspire.MongoDB.Driver.Tests/MongoDbContainerFixture.cs (hunk 6 above)
Comment (eerhardt) [reply tests/Aspire.MongoDB.Driver.Tests/MongoDbContainerFixture.cs:38]: > How is the container setup/started in case of the playground?

https://github.com/dotnet/aspire/blob/ded1eaef1592000b7c99136d1cc9a9085170f07b/playground/mongo/Mongo.AppHost/Program.cs#L6

File: tests/Aspire.MongoDB.Driver.Tests/MongoDbContainerFixture.cs (hunk 7 above)
Comment (radical) [reply tests/Aspire.MongoDB.Driver.Tests/MongoDbContainerFixture.cs:24]: Do the versions need to be surfaced in the api too? If it is just for tests, and wherever they are used in the repo right now, then we can use a shared `.cs` file for that.

Comment (eerhardt) [reply tests/Aspire.MongoDB.Driver.Tests/MongoDbContainerFixture.cs:24]: > Do the versions need to be surfaced in the api too?

No, I don't believe so.

> we can use a shared .cs file for that.

That sounds good to me.

cc @mitchdenny

File: tests/Aspire.MongoDB.Driver.Tests/MongoDbContainerFixture.cs (hunk 6 above)
Comment (radical) [reply tests/Aspire.MongoDB.Driver.Tests/MongoDbContainerFixture.cs:38]: If the default is to start it without any user/pass then we can do the same in tests too.
```csharp
            Container = new MongoDbBuilder()
                .WithImage("mongo:7.0.5")
                .WithUsername(null)
                .WithPassword(null)
                .Build();
```

**Review by eerhardt — APPROVED** [review]
Nice! This looks really clean. LGTM

File: tests/Aspire.MongoDB.Driver.Tests/MongoDbContainerFixture.cs (hunk 7 above)
Comment (radical) [reply tests/Aspire.MongoDB.Driver.Tests/MongoDbContainerFixture.cs:24]: I'll cover this in a follow up PR.
